Serialization issues as soon as Grid added to application

**Setup:**
Java 12.0.1, Maven 3.6.1, MacOS 10.14.5, Intellij IDEA 2019.1

**Steps:**
1. Pull RC1 Spring Boot starter from https://vaadin.com/start/pre-release/project-base
2. Start it with `mvn spring-boot:start`
3. Verify application has started
4. Add Grid to MainView class `Grid<String> grid = new Grid<>(); add(grid);`
5. Rebuild project to cause spring dev tools to restart app

**Expected outcome:**
Application starts without exceptions

**Actual outcome:**
Grid brings in some classes that are serializable? And this causes exceptions when application starts.

```
2019-06-25 13:36:05.551 ERROR 20856 --- [  restartedMain] o.a.catalina.session.StandardManager     : Exception loading sessions from persistent storage

java.io.InvalidObjectException: ReflectiveOperationException during deserialization
	at java.base/java.lang.invoke.SerializedLambda.readResolve(SerializedLambda.java:241) ~[na:na]
	at java.base/jdk.internal.reflect.GeneratedMethodAccessor65.invoke(Unknown Source) ~[na:na]
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43) ~[na:na]
	at java.base/java.lang.reflect.Method.invoke(Method.java:567) ~[na:na]
...

Caused by: java.lang.reflect.InvocationTargetException: null
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method) ~[na:na]
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62) ~[na:na]
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43) ~[na:na]
	at java.base/java.lang.reflect.Method.invoke(Method.java:567) ~[na:na]
	at java.base/java.lang.invoke.SerializedLambda.readResolve(SerializedLambda.java:239) ~[na:na]
	... 217 common frames omitted
Caused by: java.lang.ClassCastException: class java.lang.invoke.SerializedLambda cannot be cast to class com.vaadin.flow.dom.PropertyChangeListener (java.lang.invoke.SerializedLambda is in module java.base of loader 'bootstrap'; com.vaadin.flow.dom.PropertyChangeListener is in unnamed module of loader 'app')
	at com.vaadin.flow.internal.nodefeature.ElementPropertyMap.$deserializeLambda$(ElementPropertyMap.java:48) ~[flow-server-2.0.0.jar:2.0.0]
	... 222 common frames omitted
```

UPDATE: Exception is different now in latest RC versions, see comments!

Can't replicate by running the project, but with the junit test
```

    @Test
    public void testLayoutSerialization() throws IOException, ClassNotFoundException {
            final Div root = new Div();
            root.add(new Input());

            final ByteArrayOutputStream bout = new ByteArrayOutputStream();
            try (ObjectOutputStream oout = new ObjectOutputStream(bout))
            {
                oout.writeObject(root);
            }
            try (ObjectInputStream in = new ObjectInputStream(
                    new ByteArrayInputStream(bout.toByteArray())))
            {
                in.readObject();
            }
    }
```

it reliably fails with 
```
Caused by: java.lang.ClassCastException: java.lang.invoke.SerializedLambda cannot be cast to com.vaadin.flow.dom.PropertyChangeListener
	at com.vaadin.flow.internal.nodefeature.ElementPropertyMap.$deserializeLambda$(ElementPropertyMap.java:48)
	... 146 more
```

--- a/flow-html-components/pom.xml
+++ b/flow-html-components/pom.xml
@@ -1,5 +1,6 @@
 <?xml version="1.0" encoding="UTF-8"?>
-<project xmlns="http://maven.apache.org/POM/4.0.0" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
+<project xmlns="http://maven.apache.org/POM/4.0.0"
+    xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
     xsi:schemaLocation="http://maven.apache.org/POM/4.0.0 http://maven.apache.org/xsd/maven-4.0.0.xsd">
     <modelVersion>4.0.0</modelVersion>
     <parent>
@@ -31,6 +32,12 @@
             <scope>provided</scope>
         </dependency>
 
+        <dependency>
+            <groupId>com.vaadin</groupId>
+            <artifactId>flow-test-generic</artifactId>
+            <version>2.0-SNAPSHOT</version>
+            <scope>test</scope>
+        </dependency>
     </dependencies>
 
     <build>

--- a/flow-html-components/pom.xml
+++ b/flow-html-components/pom.xml
@@ -35,7 +35,7 @@
         <dependency>
             <groupId>com.vaadin</groupId>
             <artifactId>flow-test-generic</artifactId>
-            <version>2.0-SNAPSHOT</version>
+            <version>${project.version}</version>
             <scope>test</scope>
         </dependency>
     </dependencies>

--- a/flow-html-components/src/test/java/com/vaadin/flow/component/html/HtmlComponentSerializableTest.java
+++ b/flow-html-components/src/test/java/com/vaadin/flow/component/html/HtmlComponentSerializableTest.java
@@ -1,0 +1,22 @@
+/*
+ * Copyright 2000-2018 Vaadin Ltd.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not
+ * use this file except in compliance with the License. You may obtain a copy of
+ * the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations under
+ * the License.
+ */
+package com.vaadin.flow.component.html;
+
+import com.vaadin.flow.testutil.ClassesSerializableTest;
+
+public class HtmlComponentSerializableTest extends ClassesSerializableTest {
+
+}

--- a/flow-server/src/main/java/com/vaadin/flow/component/AbstractSinglePropertyField.java
+++ b/flow-server/src/main/java/com/vaadin/flow/component/AbstractSinglePropertyField.java
@@ -22,6 +22,7 @@ import java.util.Objects;
 import java.util.stream.Collectors;
 
 import com.googlecode.gentyref.GenericTypeReflector;
+
 import com.vaadin.flow.dom.DomListenerRegistration;
 import com.vaadin.flow.dom.Element;
 import com.vaadin.flow.dom.PropertyChangeEvent;
@@ -272,12 +273,13 @@ public class AbstractSinglePropertyField<C extends AbstractField<C, T>, T>
 
     /**
      * Sets the name of the DOM event for which property values are synchronized
-     * from the client to the server. By default, the event name is the
-     * property name with <code>-changed</code> appended. This means that if the
-     * property name is <code>value</code>, then the event default name is
+     * from the client to the server. By default, the event name is the property
+     * name with <code>-changed</code> appended. This means that if the property
+     * name is <code>value</code>, then the event default name is
      * <code>value-changed</code>.
      *
-     * @see Element#addPropertyChangeListener(String, String, PropertyChangeListener)
+     * @see Element#addPropertyChangeListener(String, String,
+     *      PropertyChangeListener)
      * @see #getSynchronizationRegistration()
      *
      * @param synchronizedEvent
@@ -293,9 +295,9 @@ public class AbstractSinglePropertyField<C extends AbstractField<C, T>, T>
      * property value. The registration is created by
      * {@link #setSynchronizedEvent(String)}.
      *
-     * @return the registration of the DOM event listener that synchronizes
-     *         the property value, or <code>null</code>
-     *         if property synchronization is disabled
+     * @return the registration of the DOM event listener that synchronizes the
+     *         property value, or <code>null</code> if property synchronization
+     *         is disabled
      */
     protected DomListenerRegistration getSynchronizationRegistration() {
         return synchronizationRegistration;
@@ -308,10 +310,16 @@ public class AbstractSinglePropertyField<C extends AbstractField<C, T>, T>
             synchronizationRegistration.remove();
         }
         if (propChangeEvent != null) {
-            synchronizationRegistration = getElement().addPropertyChangeListener(
-                    propertyName,
-                    propChangeEvent,
-                    this::handlePropertyChange);
+            synchronizationRegistration = getElement()
+                    .addPropertyChangeListener(propertyName, propChangeEvent,
+                            new PropertyChangeListener() {
+
+                                @Override
+                                public void propertyChange(
+                                        PropertyChangeEvent event) {
+                                    handlePropertyChange(event);
+                                }
+                            });
         } else {
             synchronizationRegistration = null;
         }

--- a/flow-server/src/main/java/com/vaadin/flow/component/AbstractSinglePropertyField.java
+++ b/flow-server/src/main/java/com/vaadin/flow/component/AbstractSinglePropertyField.java
@@ -312,6 +312,11 @@ public class AbstractSinglePropertyField<C extends AbstractField<C, T>, T>
         if (propChangeEvent != null) {
             synchronizationRegistration = getElement()
                     .addPropertyChangeListener(propertyName, propChangeEvent,
+                            // This explicit class instantiation is the
+                            // workaround
+                            // which fixes a JVM optimization+serialization bug.
+                            // Do not convert to lambda.
+                            // See #5973.
                             new PropertyChangeListener() {
 
                                 @Override

--- a/flow-test-generic/src/main/java/com/vaadin/flow/testutil/ClassesSerializableTest.java
+++ b/flow-test-generic/src/main/java/com/vaadin/flow/testutil/ClassesSerializableTest.java
@@ -25,6 +25,7 @@ import java.io.ObjectOutputStream;
 import java.io.Serializable;
 import java.lang.reflect.Constructor;
 import java.lang.reflect.Field;
+import java.lang.reflect.InvocationTargetException;
 import java.lang.reflect.Method;
 import java.lang.reflect.Modifier;
 import java.lang.reflect.Type;
@@ -61,16 +62,25 @@ public abstract class ClassesSerializableTest {
 
     private final Logger logger = LoggerFactory.getLogger(getClass());
 
+    private final Class<?> COMPONENT_CLASS = loadComponent(
+            "com.vaadin.flow.component.Component");
+    private final Class<?> DIV_CLASS = loadComponent(
+            "com.vaadin.flow.component.html.Div");
+
+    private final Class<?> UI_CLASS = loadComponent(
+            "com.vaadin.flow.component.UI");
+
     @SuppressWarnings("WeakerAccess")
     protected Stream<String> getExcludedPatterns() {
         return Stream.of(
                 "com\\.vaadin\\.flow\\.data\\.validator\\.BeanValidator\\$LazyFactoryInitializer",
                 "com\\.vaadin\\.flow\\.internal\\.BeanUtil\\$LazyValidationAvailability",
-                ".*\\.slf4j\\..*",
-                ".*\\.testbench\\..*",
-                ".*\\.testutil\\..*",
-                ".*\\.demo\\..*",
-                "com\\.vaadin\\..*Util(s)?(\\$\\w+)?$", //Various utils with inner classes
+                ".*\\.slf4j\\..*", ".*\\.testbench\\..*", ".*\\.testutil\\..*",
+                ".*\\.demo\\..*", "com\\.vaadin\\..*Util(s)?(\\$\\w+)?$", // Various
+                                                                          // utils
+                                                                          // with
+                                                                          // inner
+                                                                          // classes
 
                 "com\\.vaadin\\.flow\\.data\\.provider\\.InMemoryDataProviderHelpers",
                 "com\\.vaadin\\.flow\\.dom\\.ElementConstants",
@@ -127,8 +137,12 @@ public abstract class ClassesSerializableTest {
                 "com\\.vaadin\\.flow\\.server\\.VaadinRequest",
                 "com\\.vaadin\\.flow\\.router\\.RouteNotFoundError\\$LazyInit",
                 "com\\.vaadin\\.flow\\.component\\.polymertemplate\\.TemplateDataAnalyzer\\$.*",
-                "com\\.vaadin\\.flow\\.component\\.HtmlComponent",// De-facto abstract class
-                "com\\.vaadin\\.flow\\.component\\.HtmlContainer",// De-facto abstract class
+                "com\\.vaadin\\.flow\\.component\\.HtmlComponent", // De-facto
+                                                                   // abstract
+                                                                   // class
+                "com\\.vaadin\\.flow\\.component\\.HtmlContainer", // De-facto
+                                                                   // abstract
+                                                                   // class
                 "com\\.vaadin\\.flow\\.component\\.polymertemplate\\.TemplateInitializer(\\$.*)?",
                 "com\\.vaadin\\.flow\\.component\\.polymertemplate\\.TemplateParser(\\$.*)?",
                 "com\\.vaadin\\.flow\\.dom\\.impl\\.ThemeListImpl\\$ThemeListIterator",
@@ -157,9 +171,8 @@ public abstract class ClassesSerializableTest {
                 "com\\.vaadin\\.flow\\.server\\.frontend\\.scanner\\..*",
                 "com\\.vaadin\\.flow\\.server\\.frontend\\.JarContentsManager",
 
-                //Various test classes
-                ".*\\.test(s)?\\..*",
-                ".*Test.*",
+                // Various test classes
+                ".*\\.test(s)?\\..*", ".*Test.*",
                 "com\\.vaadin\\.flow\\.server\\.MockVaadinServletService",
                 "com\\.vaadin\\.flow\\.server\\.MockServletServiceSessionSetup",
                 "com\\.vaadin\\.flow\\.server\\.MockServletConfig",
@@ -354,6 +367,7 @@ public abstract class ClassesSerializableTest {
             if (!cls.isInterface()
                     && !Modifier.isAbstract(cls.getModifiers())) {
                 serializeAndDeserialize(cls);
+                serializeAndDeserializeInsideContainer(cls);
             }
 
             // report non-serializable classes and interfaces
@@ -377,23 +391,66 @@ public abstract class ClassesSerializableTest {
 
     private void serializeAndDeserialize(Class<?> clazz) {
         try {
-            Optional<Constructor<?>> defaultCtor = Stream
-                    .of(clazz.getDeclaredConstructors())
-                    .filter(ctor -> ctor.getParameterCount() == 0).findFirst();
-            if (!defaultCtor.isPresent()) {
-                return;
-            }
-            defaultCtor.get().setAccessible(true);
-            setupThreadLocals();
-            Object instance;
-            try {
-                instance = defaultCtor.get().newInstance();
-            } finally {
-                resetThreadLocals();
-            }
+            Object instance = instantiate(clazz);
             serializeAndDeserialize(instance);
         } catch (Throwable e) {
             throw new AssertionError(clazz.getName(), e);
+        }
+    }
+
+    private void serializeAndDeserializeInsideContainer(Class<?> clazz) {
+        try {
+            if (DIV_CLASS == null || COMPONENT_CLASS == null) {
+                return;
+            }
+            if (!COMPONENT_CLASS.isAssignableFrom(clazz)) {
+                return;
+            }
+            if (UI_CLASS != null && UI_CLASS.isAssignableFrom(clazz)) {
+                return;
+            }
+            Object div = instantiate(DIV_CLASS);
+            Object instance = instantiate(clazz);
+            if (instance == null) {
+                return;
+            }
+
+            Object divElement = getElement(div);
+            Optional<Method> setChild = Stream
+                    .of(divElement.getClass().getMethods())
+                    .filter(method -> "setChild".equals(method.getName()))
+                    .findFirst();
+            setChild.get().invoke(divElement, 0, getElement(instance));
+            serializeAndDeserialize(div);
+        } catch (
+
+        Throwable e) {
+            throw new AssertionError(clazz.getName(), e);
+        }
+    }
+
+    private Object getElement(Object obj) throws NoSuchMethodException,
+            SecurityException, IllegalAccessException, IllegalArgumentException,
+            InvocationTargetException {
+        Method method = obj.getClass().getMethod("getElement");
+        return method.invoke(obj);
+    }
+
+    private Object instantiate(Class<?> clazz)
+            throws InstantiationException, IllegalAccessException,
+            IllegalArgumentException, InvocationTargetException {
+        Optional<Constructor<?>> defaultCtor = Stream
+                .of(clazz.getDeclaredConstructors())
+                .filter(ctor -> ctor.getParameterCount() == 0).findFirst();
+        if (!defaultCtor.isPresent()) {
+            return null;
+        }
+        defaultCtor.get().setAccessible(true);
+        setupThreadLocals();
+        try {
+            return defaultCtor.get().newInstance();
+        } finally {
+            resetThreadLocals();
         }
     }
 
@@ -495,6 +552,14 @@ public abstract class ClassesSerializableTest {
             }
         }
         return classes;
+    }
+
+    private Class<?> loadComponent(String fqn) {
+        try {
+            return Class.forName(fqn);
+        } catch (ClassNotFoundException e) {
+            return null;
+        }
     }
 
 }

--- a/flow-test-generic/src/main/java/com/vaadin/flow/testutil/ClassesSerializableTest.java
+++ b/flow-test-generic/src/main/java/com/vaadin/flow/testutil/ClassesSerializableTest.java
@@ -76,11 +76,8 @@ public abstract class ClassesSerializableTest {
                 "com\\.vaadin\\.flow\\.data\\.validator\\.BeanValidator\\$LazyFactoryInitializer",
                 "com\\.vaadin\\.flow\\.internal\\.BeanUtil\\$LazyValidationAvailability",
                 ".*\\.slf4j\\..*", ".*\\.testbench\\..*", ".*\\.testutil\\..*",
-                ".*\\.demo\\..*", "com\\.vaadin\\..*Util(s)?(\\$\\w+)?$", // Various
-                                                                          // utils
-                                                                          // with
-                                                                          // inner
-                                                                          // classes
+                // Various utils with inner classes
+                ".*\\.demo\\..*", "com\\.vaadin\\..*Util(s)?(\\$\\w+)?$",
 
                 "com\\.vaadin\\.flow\\.data\\.provider\\.InMemoryDataProviderHelpers",
                 "com\\.vaadin\\.flow\\.dom\\.ElementConstants",
@@ -137,12 +134,10 @@ public abstract class ClassesSerializableTest {
                 "com\\.vaadin\\.flow\\.server\\.VaadinRequest",
                 "com\\.vaadin\\.flow\\.router\\.RouteNotFoundError\\$LazyInit",
                 "com\\.vaadin\\.flow\\.component\\.polymertemplate\\.TemplateDataAnalyzer\\$.*",
-                "com\\.vaadin\\.flow\\.component\\.HtmlComponent", // De-facto
-                                                                   // abstract
-                                                                   // class
-                "com\\.vaadin\\.flow\\.component\\.HtmlContainer", // De-facto
-                                                                   // abstract
-                                                                   // class
+                // De-facto abstract class
+                "com\\.vaadin\\.flow\\.component\\.HtmlComponent",
+                // De-facto abstract class
+                "com\\.vaadin\\.flow\\.component\\.HtmlContainer",
                 "com\\.vaadin\\.flow\\.component\\.polymertemplate\\.TemplateInitializer(\\$.*)?",
                 "com\\.vaadin\\.flow\\.component\\.polymertemplate\\.TemplateParser(\\$.*)?",
                 "com\\.vaadin\\.flow\\.dom\\.impl\\.ThemeListImpl\\$ThemeListIterator",
@@ -422,9 +417,7 @@ public abstract class ClassesSerializableTest {
                     .findFirst();
             setChild.get().invoke(divElement, 0, getElement(instance));
             serializeAndDeserialize(div);
-        } catch (
-
-        Throwable e) {
+        } catch (Throwable e) {
             throw new AssertionError(clazz.getName(), e);
         }
     }
@@ -436,9 +429,8 @@ public abstract class ClassesSerializableTest {
         return method.invoke(obj);
     }
 
-    private Object instantiate(Class<?> clazz)
-            throws InstantiationException, IllegalAccessException,
-            IllegalArgumentException, InvocationTargetException {
+    private Object instantiate(Class<?> clazz) throws InstantiationException,
+            IllegalAccessException, InvocationTargetException {
         Optional<Constructor<?>> defaultCtor = Stream
                 .of(clazz.getDeclaredConstructors())
                 .filter(ctor -> ctor.getParameterCount() == 0).findFirst();

